OfflineTileListRegionDefinition

JesseCrocker opened the pull request.

mcglincy commented on include/mbgl/storage/offline.hpp:
```diff
@@ -62,10 +63,34 @@ class OfflineGeometryRegionDefinition {
+public:
+  OfflineTileListRegionDefinition(std::string, LatLngBounds, double, double, float, std::vector<CanonicalTileID>);
+  
+  /* Private */
```
out of curiosity: what's up with the comment "/* Private */" as opposed to actually making the members private (with the private: keyword)? I see that comment is used elsewhere.

JesseCrocker replied:
no idea.

mcglincy commented on include/mbgl/storage/offline.hpp:
```diff
@@ -62,10 +63,34 @@ class OfflineGeometryRegionDefinition {
 };
 
+/*
+ * An offline region defined by a style URL, list of tiles, and
```
"list of tile IDs", for clarity ?

mcglincy commented on platform/darwin/src/MGLOfflinePack.mm:
```diff
@@ -4,6 +4,8 @@
 #import "MGLOfflineRegion_Private.h"
 #import "MGLTilePyramidOfflineRegion.h"
 #import "MGLTilePyramidOfflineRegion_Private.h"
+#import "MGLTileListOfflineRegion.h"
```
I was going to nit about alphabetizing the imports, but they're already screwed up :P

mcglincy commented on platform/darwin/src/MGLTileID.h:
```diff
@@ -0,0 +1,19 @@
+} MGLTileID;
+
+// Returns an MGTileID tile from tile key
+FOUNDATION_EXTERN MGL_EXPORT MGLTileID MGLTileIDFromKey(uint64_t tilekey);
```
nit: camel case "tileKey" var name?

mcglincy commented on platform/darwin/src/MGLTileID.h:
```diff
@@ -0,0 +1,19 @@
+
+typedef struct
+{
+  uint8_t z;
```
is z actually zoom? if so, use "zoom" as the var name?

JesseCrocker replied:
This is following a convention set in the c++ equivalent of this class `CanonicalTileID`.

mcglincy commented on platform/darwin/src/MGLTileID.m:
```diff
@@ -0,0 +1,42 @@
+//  ios
+//
+//  Created by Anna Johnson on 2/7/17.
+//  Copyright © 2017 Mapbox. All rights reserved.
```
what's up with this 2017 copyright?

JesseCrocker replied:
Anna originally wrote this code 2 years ago, but mapbox-gl has changed so much since then that the originally branches were totally un-mergeable, so the code in this PR is not original, it's copy pasted, though half of this is actually new code.

mcglincy commented on platform/darwin/src/MGLTileID.h:
```diff
@@ -0,0 +1,19 @@
+#import <Foundation/Foundation.h>
```
If we intend to upstream this code, maybe add "/* */" style comments and description to make this code look like other .h/.m in https://github.com/trailbehind/mapbox-gl-native/tree/master/platform/darwin/src

mcglincy commented on platform/darwin/src/MGLTileID.m:
```diff
@@ -0,0 +1,42 @@
+    return t;
+}
+
+/// Returns a unique key of the tile for use in the SQLite cache
```
nit: "///"

mcglincy commented on platform/darwin/src/MGLTileID.h:
```diff
@@ -0,0 +1,19 @@
+// Returns a unique key of the tile
+FOUNDATION_EXTERN MGL_EXPORT uint64_t MGLTileKey(MGLTileID tile);
+
+FOUNDATION_EXTERN MGL_EXPORT MGLTileID MGLTileIDMake(uint32_t x, uint32_t y, uint8_t z);
```
nit: should our parameter order by z[oom], x, y... like the struct?

mcglincy commented on platform/darwin/src/MGLTileID.m:
```diff
@@ -0,0 +1,42 @@
+}
+
+
+MGLTileID MGLTileIDMake(uint32_t x, uint32_t y, uint8_t z) {
```
replace MGLTileIDMake() with a constructor on the struct itself?

(unless we're also intending to use aggregate initialization for the struct, too?)

mcglincy replied:
nm, didn't see this is just c, and not c++...

mcglincy commented on platform/darwin/src/MGLTileListOfflineRegion.h:
```diff
@@ -0,0 +1,83 @@
+/**
+ A specific list of tileIds to download for this region
+ */
+@property (nonatomic, readonly, nullable) NSArray* tileList;
```
nit for pointer * placement: "NSArray *tileList"

mcglincy commented on platform/darwin/src/MGLTileListOfflineRegion.mm:
```diff
@@ -0,0 +1,162 @@
+    NSURL *_styleURL;
+}
+
+@synthesize styleURL = _styleURL;
```
I'm not familiar with non-iphone objective-c, but @synthesize should be automatic in later xcode versions (and generate _styleURL automatically)

JesseCrocker replied:
It has to be synthesized here because the @property is declared in a protocol

mcglincy commented on platform/darwin/src/MGLTileListOfflineRegion.mm:
```diff
@@ -0,0 +1,162 @@
+- (instancetype)init {
+    MGLLogInfo(@"Calling this initializer is not allowed.");
+    [NSException raise:NSGenericException format:
+     @"-[MGLTilePyramidOfflineRegion init] is unavailable. "
```
copy-paste error with "MGLTilePyramidOfflineRegion"?

mcglincy commented on platform/darwin/src/MGLTileListOfflineRegion.h:
```diff
@@ -0,0 +1,83 @@
+    level.
+ @params tileList list of tiles
+ */
+- (instancetype)initWithStyleURL:(NSURL *)styleURL bounds:(MGLCoordinateBounds)bounds fromZoomLevel:(double)minimumZoomLevel toZoomLevel:(double)maximumZoomLevel tileList:(nullable NSArray*)tileList;
```
nit: "nullable NSArray*" => "nullable NSArray *" (add space between type and *)

mcglincy commented on platform/darwin/src/MGLTileListOfflineRegion.h:
```diff
@@ -0,0 +1,83 @@
+    parameter is set to a value greater than the tile source’s minimum zoom
+    level, the download covers zoom levels up to the tile source’s maximum zoom
+    level.
+ @params tileList list of tiles
```
maybe mention the type of the list in the comment?

mcglincy commented on platform/darwin/src/MGLTileListOfflineRegion.mm:
```diff
@@ -0,0 +1,162 @@
+            styleURL = [MGLStyle streetsStyleURLWithVersion:MGLStyleDefaultVersion];
+        }
+
+        if (!styleURL.scheme) {
```
does streetsStyleURLWithVersion possible return a scheme-less URL that we need to check, too? If not, this block might be clearer as:

if (!styleURL) {
 ...
} else if (!styleURL.scheme) {
 ...
}

mcglincy commented on platform/darwin/src/MGLTileListOfflineRegion.mm:
```diff
@@ -0,0 +1,162 @@
+    MGLCoordinateBounds bounds = MGLCoordinateBoundsFromLatLngBounds(definition.bounds);
+    NSMutableArray *tileList = [NSMutableArray arrayWithCapacity:definition.tileList.size()];
+    for(mbgl::CanonicalTileID t: definition.tileList) {
+        NSNumber *tileIdNumber = @(MGLTileKey(MGLTileIDMake(t.x, t.y, t.z)));
```
nit: "tileIdNumber" is really a "tileKey".

mcglincy commented on platform/darwin/src/MGLTileListOfflineRegion.mm:
```diff
@@ -0,0 +1,162 @@
+    NSURL *styleURL = [NSURL URLWithString:@(definition.styleURL.c_str())];
+    MGLCoordinateBounds bounds = MGLCoordinateBoundsFromLatLngBounds(definition.bounds);
+    NSMutableArray *tileList = [NSMutableArray arrayWithCapacity:definition.tileList.size()];
+    for(mbgl::CanonicalTileID t: definition.tileList) {
```
nit: space before and after ":" in for loop syntax

mcglincy commented on platform/darwin/src/MGLTileListOfflineRegion.mm:
```diff
@@ -0,0 +1,162 @@
+- (instancetype)initWithOfflineRegionDefinition:(const mbgl::OfflineTileListRegionDefinition &)definition {
+    NSURL *styleURL = [NSURL URLWithString:@(definition.styleURL.c_str())];
+    MGLCoordinateBounds bounds = MGLCoordinateBoundsFromLatLngBounds(definition.bounds);
+    NSMutableArray *tileList = [NSMutableArray arrayWithCapacity:definition.tileList.size()];
```
nit: rename "tileList" to "tileKeys" ?

mcglincy commented on platform/darwin/src/MGLTileListOfflineRegion.mm:
```diff
@@ -0,0 +1,162 @@
+#endif
+    std::vector<mbgl::CanonicalTileID> tileVector;
+    tileVector.reserve(_tileList.count);
+    for (NSNumber* value in _tileList) {
```
nit: "NSNumber *value"

POINTER NITS AHOY!

mcglincy commented on platform/darwin/src/MGLTileListOfflineRegion.mm:
```diff
@@ -0,0 +1,162 @@
+    tileVector.reserve(_tileList.count);
+    for (NSNumber* value in _tileList) {
+        MGLTileID tileId = MGLTileIDFromKey([value unsignedLongLongValue]);
+        mbgl::CanonicalTileID canonicalTileID = mbgl::CanonicalTileID(tileId.z, tileId.x,tileId.y);
```
nit: missing space in "tileId.x,tileId.y"

mcglincy commented on platform/darwin/src/MGLTileListOfflineRegion.mm:
```diff
@@ -0,0 +1,162 @@
+}
+
+- (void)encodeWithCoder:(NSCoder *)coder
+{
```
nit: move { to end of previous line

mcglincy commented on platform/default/mbgl/storage/offline.cpp:
```diff
@@ -120,6 +151,23 @@ std::string encodeOfflineRegionDefinition(const OfflineRegionDefinition& region)
+                bounds.PushBack(_region.bounds.east(), doc.GetAllocator());
+                doc.AddMember("bounds", bounds, doc.GetAllocator());
+                rapidjson::GenericValue<rapidjson::UTF8<>, rapidjson::CrtAllocator> tileList(rapidjson::kArrayType);
+                for ( auto &i : _region.tileList ) {
```
maybe choose a variable besides "i" for iteration, which usually implies a simple int counter.

mcglincy commented on platform/default/mbgl/storage/offline_download.cpp:
```diff
@@ -57,7 +57,14 @@ void tileCover(const OfflineRegionDefinition& definition, style::SourceType type
+                [&](const OfflineGeometryRegionDefinition& reg){ tileCover(reg.geometry, z, fn); },
+                [&](const OfflineTileListRegionDefinition& reg){
+                    if(z == clampedZoomRange.min) {
+                      for(CanonicalTileID t: reg.tileList) {
```
nit: space to either side of :

mcglincy commented on platform/default/mbgl/storage/offline_download.cpp:
```diff
@@ -57,7 +57,14 @@ void tileCover(const OfflineRegionDefinition& definition, style::SourceType type
-                [&](const OfflineGeometryRegionDefinition& reg){ tileCover(reg.geometry, z, fn); }
+                [&](const OfflineGeometryRegionDefinition& reg){ tileCover(reg.geometry, z, fn); },
+                [&](const OfflineTileListRegionDefinition& reg){
+                    if(z == clampedZoomRange.min) {
```
I'm new to mapbox / tile internals - can you help me understand why we apply the function to each tile when we're at min zoom?

JesseCrocker replied:
fn is actually just a function that adds an item to a an std::vector. The way this is written is totally dumb, it doesn't matter what the zoom is, just that it only happens one time. It would make more sense if this just wasnt in this zoom loop, but variant.match requires that all of the variant types have an entry.

mcglincy commented:
Bunch o' nits, but code generally LGTM. 

I don't fully understand the problem space (mapbox, tiles), so I defer to you that we actually want/need to do this :)

Commit:
OfflineTileListRegionDefinition

Squashed commit of the following:

commit f435f0cfdfc5a1a99af959304a2e236f43b065a1
Author: Jesse Crocker <jesse@gaiagps.com>
Date:   Wed Feb 20 07:46:47 2019 -0700

    update for iOS 4.8 release

commit bc3aca4756c6ed8b7e3dc7eb64af43ede6ce9231
Author: Jesse Crocker <jesse@gaiagps.com>
Date:   Wed Feb 20 06:53:33 2019 -0700

    OfflineTileListRegionDefinition
